app/prepare_proposal.go

```diff
@@ -42,38 +35,45 @@ func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePr
 		ante.DefaultSigVerificationGasConsumer,
 		app.IBCKeeper,
 		app.ParamsKeeper,
-		app.MsgGateKeeper,
+		&app.CircuitKeeper,
+		app.BlockedParamsGovernance(),
 	)
 
 	// Filter out invalid transactions.
-	txs := FilterTxs(app.Logger(), sdkCtx, handler, app.txConfig, req.BlockData.Txs)
+	txs := FilterTxs(app.Logger(), ctx, handler, app.encodingConfig.TxConfig, req.Txs)
 
 	// Build the square from the set of valid and prioritised transactions.
 	// The txs returned are the ones used in the square and block.
 	var (
 		dataSquareBytes [][]byte
-		err             error
 		size            uint64
+		err             error
 	)
-	switch app.AppVersion() {
-	case v3:
+
+	appVersion, err := app.AppVersion(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	switch appVersion {
+	case v4, v3:
 		var dataSquare squarev2.Square
 		dataSquare, txs, err = squarev2.Build(txs,
-			app.MaxEffectiveSquareSize(sdkCtx),
-			appconsts.SubtreeRootThreshold(app.GetBaseApp().AppVersion()),
+			app.MaxEffectiveSquareSize(ctx),
+			appconsts.SubtreeRootThreshold(appVersion),
 		)
 		dataSquareBytes = sharev2.ToBytes(dataSquare)
 		size = uint64(dataSquare.Size())
 	case v2, v1:
 		var dataSquare square.Square
 		dataSquare, txs, err = square.Build(txs,
-			app.MaxEffectiveSquareSize(sdkCtx),
-			appconsts.SubtreeRootThreshold(app.GetBaseApp().AppVersion()),
+			app.MaxEffectiveSquareSize(ctx),
+			appconsts.SubtreeRootThreshold(appVersion),
```

Comment:
do we need this switch at all? my understanding is that the multiplexer handles this

Reply:
agreed, my understanding is that all of this switching and conditionals on AppVersion can be removed now. It's just a matter of combing through the remainder of files. I believe there's still some in tests also. I'll create an issue on this and get started on it before end of week

Reply:
dope, thanks!

Reply:
https://github.com/01builders/celestia-app/pull/47 looks so much cleaner! 🙏🏻